### ISSUE TITLE
TypedDict: `__required_keys__` and `__optional_keys__` can be wrong in the presence of inheritance

### ISSUE THREAD

**Issue by JelleZijlstra** [python/cpython#112509]
# Bug report

### Bug description:

```python
from typing import NotRequired, Required, TypedDict

class A(TypedDict):
    a: NotRequired[int]
    b: Required[int]

class B(A):
    a: Required[int]
    b: NotRequired[int]

print(B.__required_keys__)
print(B.__optional_keys__)
```

This will print (tried on 3.11 and current-ish main):
```
frozenset({'b', 'a'})
frozenset({'b', 'a'})
```

But obviously, a single key should only be either Required or NotRequired, not both. The child class's version should prevail.

@alicederyn and I discovered this while discussing the implementation of PEP-705 (https://github.com/python/typing_extensions/pull/284/#discussion_r1408644357). cc @davidfstr for PEP 655.

I also see some problems with how type checkers handle this case, but i'll post about that separately.

### CPython versions tested on:

3.11, CPython main branch

### Operating systems tested on:

macOS

### Linked PRs
* gh-112512
* gh-112530
* gh-112531

**Comment by alicederyn** [comment]
This example is illegal according to PEP-589.

> A TypedDict type with a required key is not consistent with a TypedDict type where the same key is a non-required key, since the latter allows keys to be deleted.

Concretely, B cannot be a subclass of A as it "removes" the ability to delete key 'b' (breaking substitutability) and also adds None to the possible values returned by `get('a')` (also breaking substitutability).

**Comment by alicederyn** [comment]
> If we had prohibited that at runtime, it would make it harder for people to use PEP 705 on older Python versions.

I wasn't suggesting prohibiting it at runtime -- just that changing the runtime behaviour of an unsupported construction seems unnecessary.

Using PEP-705 with a regular TypedDict instead of importing it from typing_extensions will give the wrong behaviour at runtime anyway, so I don't think this change will benefit it.

**Comment by AlexWaygood** [comment]
This is fixed at runtime now, and backported to 3.11+, so I'll close this out now

**Comment by JelleZijlstra** [comment]
@alicederyn that's a fair point, but I think the behavior here (the same key being in both required and optional keys) is quite clearly buggy; there is room for disagreement (or future spec changes) in the type checker behavior; and the fix was simple.
